server/src/main/java/io/druid/server/coordinator/DruidCoordinator.java

```diff
@@ -643,9 +644,51 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
       );
     }
 
+    if (config.isKillSegments()) {
+      helpers.add(new DruidCoordinatorSegmentKiller(indexingServiceClient, config.getCoordinatorKillPeriod(), whitelistRef));
+    }
+
     return ImmutableList.copyOf(helpers);
   }
 
+  static class DruidCoordinatorSegmentKiller implements DruidCoordinatorHelper
+  {
+    private static final Interval UBER_INTERVAL = Interval.parse("1970/3000");
+
+    private final long period;
+    private long lastKillTime = 0;
+
+    private final IndexingServiceClient indexingServiceClient;
+    private final AtomicReference<DatasourceWhitelist> whitelistRef;
+
+    DruidCoordinatorSegmentKiller(
+        IndexingServiceClient indexingServiceClient,
+        Duration period,
+        AtomicReference<DatasourceWhitelist> whitelistRef
+    )
+    {
+      this.period = period.getMillis();
+      this.indexingServiceClient = indexingServiceClient;
+      this.whitelistRef = whitelistRef;
+    }
+
+    @Override
+    public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
+    {
+      if ((lastKillTime + period) < System.currentTimeMillis()) {
+        lastKillTime = System.currentTimeMillis();
+
+        DatasourceWhitelist whitelist = whitelistRef.get();
+        for (DruidDataSource dataSource : params.getDataSources()) {
+          if (whitelist == null || whitelist.contains(dataSource.getName())) {
+            indexingServiceClient.killSegments(dataSource.getName(), UBER_INTERVAL);
```

Comment:
This will block per datasource. If one fails to post for whatever reason, it will not attempt the other kill tasks. 

But the thing that calls helper.run doesn't seem to catch exceptions per helper either, so one failure in any helper will cause the others to fail. Can you catch errors from killSegments submission per loop?